server/worker/worker.go

```diff
@@ -0,0 +1,110 @@
+package worker
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+const maxRetries = 5
+
+// Job defines an interface for jobs that can be run by the Worker
+type Job interface {
+	// Name is the unique name of the job.
+	Name() string
+
+	// Run performs the actual work.
+	Run(ctx context.Context, args map[string]interface{}) error
+}
+
+// Worker runs jobs. NOT SAFE FOR CONCURRENT USE.
+type Worker struct {
+	ds  fleet.Datastore
+	log kitlog.Logger
+
+	registry map[string]Job
+}
+
+func NewWorker(ds fleet.Datastore, log kitlog.Logger) *Worker {
+	return &Worker{
+		ds:       ds,
+		log:      log,
+		registry: make(map[string]Job),
+	}
+}
+
+func (w *Worker) Register(jobs ...Job) {
+	for _, j := range jobs {
+		w.registry[j.Name()] = j
+	}
+}
+
+func (w *Worker) QueueJob(ctx context.Context, job *fleet.Job) (*fleet.Job, error) {
+	job.State = fleet.JobStateQueued
+	return w.ds.NewJob(ctx, job)
+}
+
+// ProcessJobs processes all queued jobs.
+func (w *Worker) ProcessJobs(ctx context.Context) error {
+
+	// process jobs until there are none left
+	maxNumJobs := 100
+	for {
+		jobs, err := w.ds.GetQueuedJobs(ctx, maxNumJobs)
+		if err != nil {
+			return fmt.Errorf("get jobs: %w", err)
+		}
+
+		if len(jobs) == 0 {
+			break
+		}
+
+		for _, job := range jobs {
+			log := kitlog.With(w.log, "job_id", job.ID)
+
+			level.Debug(log).Log("msg", "processing job")
+
+			err := w.processJob(ctx, job)
+			if err != nil {
+				level.Error(log).Log("msg", "job failed", "err", err)
+				job.Error = err.Error()
+				if job.Retries < maxRetries {
```

Comment:
That's probably good enough for now, but when there are few jobs, the retry would happen pretty much instantly. We should think about some backoff for retries (it can be as simple as having a "next attempt" timestamp that we set to some time in the future and the query to load queued jobs ignores those in the future). But also, and that is often a problem with "real" queues too, many failures are just unretryable - an unexpected input value that requires some changes to the processing implementation, and in those cases, we just fail as many times as we can retry, wasting processing time. There are a limited number of errors that make sense to retry (pretty much always networking "temporary" errors), we could eventually add some error inspection to see if it makes sense to retry at all. But again, that's certainly fine for now.

Reply:
My goals for the first iteration was simplicity and minimizing developer time. There is certainly lots of room for improvement here.